### PR TITLE
CRM-21425: Make Inbound Emails Editable With Configuration Option

### PR DESCRIPTION
**Pull request by MiyaNoctem**
Overview
----------------------------------------
On some business cases, it may be required to be able to edit an activity type of 'Inbound E-mail', for example, to add an observation or edit some custom fields to implement some kind of workflow. However, this is currently not possible, as the prohibition for this type of activities is hard-coded in several parts within CiviCRM.

Before
----------------------------------------
Restrictions to make 'Inbound E-mail' activities uneditable were hard-coded into several parts of the CRM, making it impossible for users to edit these types of activities.

After
----------------------------------------
A new boolean configuration option was added to CiviMail settings, so an administrator, or user with enough permissions, can enable the option to allow users to edit activities of type 'Inbound Email' to which they have access to. This new configuration option is checked everywhere the restriction was hard-coded into the CRM.

Comments
----------------------------------------
When testing edit form for Inbound E-mails, it was found that the wysiwyg editor got broken when the activity's detail field was being rendered by the template. This was caused by the stripping of alternatives from the ezComponents-parsed contents of the e-mail, which had the side effect of also stripping the html for the textarea that was later used to build the wysiwyg editor. This was fixed by moving the stripping of alternatives before the html for the textarea was being rendered, namely at the form's preprocessing.

Documentation
-----------------------------------------
Added new documentation explaining the use of the new option on user and sysadmin guides on these PR's:

https://github.com/civicrm/civicrm-sysadmin-guide/pull/59
https://github.com/civicrm/civicrm-user-guide/pull/239

---

 * [CRM-21425: Make 'Inbound E-mail' Activities Editable](https://issues.civicrm.org/jira/browse/CRM-21425)

### CODE REVIEW

**Comment by MegaphoneJon** [comment]
A reminder that PR review now requires a PR to the docs when necessary (and adding an option to the UI qualifies). If you already have one, could you please link to it?

File: CRM/Activity/Form/Activity.php [hunk 1]
```diff
@@ -499,10 +499,14 @@ public function preProcess() {
         $params = array('id' => $this->_activityId);
         CRM_Activity_BAO_Activity::retrieve($params, $this->_values);
       }
+
       $this->set('values', $this->_values);
     }
 
     if ($this->_action & CRM_Core_Action::UPDATE) {
+      // We filter out alternatives, in case this is a stored e-mail, before sending to front-end
+      $this->_values['details'] = CRM_Utils_String::stripAlternatives($this->_values['details']);
```
Comment (mlutfy) [inline CRM/Activity/Form/Activity.php]: This seems like a good idea, but it's a separate topic, right?

File: CRM/Admin/Form/Preferences/Mailing.php [hunk 2]
```diff
@@ -108,6 +108,12 @@ public function preProcess() {
           'weight' => 11,
           'description' => 'If enabled, a randomized hash key will be used to reference the mailing URL in the mailing.viewUrl token, instead of the mailing ID',
         ),
+        'allow_edit_inbound_emails' => array(
+          'html_type' => 'checkbox',
+          'title' => ts('Allow to Edit Inbound E-mails'),
+          'weight' => 12,
+          'description' => 'If enabled, users will be able to edit activities of Inbound E-mail type',
```
Comment (mlutfy) [inline CRM/Admin/Form/Preferences/Mailing.php]: Please wrap in `ts()`

**Comment by mlutfy** [comment]
This seems like a nice to have, and I don't think it's worth pushing out to an extension.

* What about default values? (install & upgrade)
* Shouldn't this register a setting under the `settings/` directory?
* I'm not sure the setting should go under "Mailing", because this is not strictly related to outbound mailings.

File: CRM/Activity/Form/Activity.php (hunk 1 above)
Comment (MiyaNoctem) [reply CRM/Activity/Form/Activity.php]: Hey, @mlutfy, thanks for the feedback. It is kind of a separate topic, but the thing is if we don't do this, the details part of the activity won't be editable, as this is currently done on the template, which breaks the html for the textarea used to show the wysiwyg editor. Let me know if you still think this should better be handled on a different PR/issue.

File: CRM/Admin/Form/Preferences/Mailing.php (hunk 2 above)
Comment (MiyaNoctem) [reply CRM/Admin/Form/Preferences/Mailing.php]: Done!

**Comment by MiyaNoctem** [comment]
@mlutfy,

Thanks again for your feedback. About your last note:

- I think leaving the new option unset as default is good, as CiviCRM will keep working as it is known to on install/upgrade, and only if you want a different behaviour, you can go and change the setting.
- I added the setting to /settings
- I kept the setting as part of Mailing since Inbound E-mail-to-activity processing is handled by CiviMail component.

**Comment by eileenmcnaughton** [comment]
Does your use case require editing the details fields (which holds the contents of the mail) - or only the surrounding fields. I'd be inclined to skip the setting and always allow editing, but to lock the one field that is a 'true record' of an incoming email

**Comment by guanhuan** [comment]
Although the details field is not directly required by our use case but I think the details does not always parse the neatest format of the email so it is likely useful to allow it to be edited as well?

The setting is probably a good practice since not everyone wants to allow the inbound mails to be edited per any other activities, instead of introducing a direct change, with the setting, we are effectively introducing an option. I feel that will be more acceptable in general.

**Comment by eileenmcnaughton** [comment]
Hmm - I guess in general we treat additional settings as a red flag that says 'we need to carefully question whether this should be in core' - I would not hesitate to make some fields editable for all on those activities (anything that does not come directly from the email). I'm still a soft minus one on adding the setting - although others may have different opinions

**Comment by guanhuan** [comment]
@eileenmcnaughton I think we can meet in the middle here and I think might make the most of sense considering your suggestion:

- allowing all other fields to be editable by default
- modifying the setting's purpose to only switch on/ off the details field editing

Please let me know your thoughts.

Best

Guanhuan

**Comment by eileenmcnaughton** [comment]
OK - go for it - I guess there is no problem other than loading core with more code to maintain to cater to a site's specific quirk or preference (or in this case it seems the possibility of one since you say it's not required for your use case)

**Comment by eileenmcnaughton** [comment]
I'd be a lot more comfortable if it were a permission than a setting TBH

**Comment by jamienovick** [comment]
Hi @guanhuan @eileenmcmaughton

I think a setting to allow editing which enables/creates a permission to also specify who can edit the details field sounds good to me.

Client only wants editing of subject field so fits their use case better also.

Thanks

J

**Comment by eileenmcnaughton** [comment]
So - we are looking at a permission instead of a setting now?

**Comment by eileenmcnaughton** [comment]
I don't understand a setting that enables a permission - ie. what that means. If it's a permission it would show up on the permissions screen & wouldn't need a setting for that

**Comment by shitijg** [comment]
Hi @eileenmcnaughton,
I have had a short discussion with @guanhuan and @jamienovick, we agree that this would be less complicated and cleaner if we just use permission instead of a setting that further activates a permission.
So we would suggest 2 new permissions here:

- Permission for editing inbound email (cannot edit content): 
CiviCRM: edit inbound email basic information
Edit all inbound email activities (for visible contacts) basic information. Content editing not allowed.


- Permission for editing inbound email including details:
CiviCRM: edit inbound email basic information and content
Edit all inbound email activities (for visible contacts) basic information and content.

In short, these are both edit permissions on top of the view permissions which we already have. They are separated out on ability to edit content. 

Let me know what you think. 

Thanks,
Shitij

**Comment by eileenmcnaughton** [comment]
I think that's OK - I would be leaving out the 'edit basic information' & assuming it - but I'm not opposed to having both

**Comment by shitijg** [comment]
Thanks for your comments @eileenmcnaughton We will update the PR with just the 2 new permissions then.

**Comment by eileenmcnaughton** [comment]
Once this is all passing can you rebase it into a single commit? The changes don't seem to be distinct enough for separate commits.

**Comment by MiyaNoctem** [comment]
Done @eileenmcnaughton, squashed everything into a single commit. Waiting for Jenkins now, but should be ok.

**Comment by MiyaNoctem** [comment]
@eileenmcnaughton , @shitijg, This seems ready now. I've updated documentation on admin and user guides with the new permissions also.

**Comment by MiyaNoctem** [comment]
@eileenmcnaughton , @colemanw ,

So you think we can remove the WIP tag from this PR and have it merged?

**Comment by eileenmcnaughton** [comment]
I think the code / approach looks good now - It needs me or someone else to do some front end testing & then is OK to merge I think

File: CRM/Activity/Form/Activity.php [hunk 3]
```diff
@@ -523,7 +523,7 @@ public function preProcess() {
 
     if ($this->_action & CRM_Core_Action::UPDATE) {
       // We filter out alternatives, in case this is a stored e-mail, before sending to front-end
-      $this->_values['details'] = CRM_Utils_String::stripAlternatives($this->_values['details']);
+      $this->_values['details'] = CRM_Utils_String::stripAlternatives($this->_values['details']) ?: '';
```
Comment (colemanw) [inline CRM/Activity/Form/Activity.php]: Huh, that's the first time I've encountered this type of expression in php. Looking it up - it's called the Elvis operator. Cool.
image

Comment (eileenmcnaughton) [reply CRM/Activity/Form/Activity.php]: lol

Comment (eileenmcnaughton) [reply CRM/Activity/Form/Activity.php]: In php7 you get a double ?? option to get around doing isset()

**Comment by eileenmcnaughton** [comment]
ug - there is a conflict (minor) now

**Comment by shitijg** [comment]
Hi @MiyaNoctem / @vinuvarshith - can you please look at the conflict(s)?

**Comment by eileenmcnaughton** [comment]
@MiyaNoctem / @vinuvarshith  I'm going to close & you can re-open when the conflict is resolved Ping me when re-opened
